<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<style type="text/css">
@font-face {
font-family: octicons-link;
src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM+8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB/aFGpk3jaTY6xa8JAGMW/O62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v+k/0an2i+itHDw3v2+9+DBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3/I7AtxEJLtzzuZfI+VVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy/Lt7Kc+0vWY/gAgIIEqAN9we0pwKXreiMasxvabDQMM4riO+qxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw+ymhce7vwM9jSqO8JyVd5RH9gyTt2+J/yUmYlIR0s04n6+7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv/ocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi+W2+MjCzMIDApSwvXzC97Z4Ig8N/BxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh/8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT+AEjAwuDFpBmA9KMDEwMCh9i/v8H8sH0/4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9/lqYwOGZxeUelN2U2R6+cArgtCJpauW7UQBqnFkUsjAY/kOU1cP+DAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl+vvmM/byA48e6tWrKArm4ZJlCbdsrxksL1AwWn/yBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO//sdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd/89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF+9JOS0nbaaYDCQfwCJ7Au3AHj+LO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm+EBXuAbHmIMSRMs+4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL+hD7C1xoaHeLJSEao0FEW14ckxC+TU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13/+lm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl+9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O/AdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB///AA8AAQAAAAAAAAAAAAAAAAABAAAAAA==) format('woff');
}
body {
-webkit-text-size-adjust: 100%;
text-size-adjust: 100%;
color: #333;
font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 16px;
line-height: 1.6;
word-wrap: break-word;
}
a {
background-color: transparent;
}
a:active,
a:hover {
outline: 0;
}
strong {
font-weight: bold;
}
h1 {
font-size: 2em;
margin: 0.67em 0;
}
img {
border: 0;
}
hr {
box-sizing: content-box;
height: 0;
}
pre {
overflow: auto;
}
code,
kbd,
pre {
font-family: monospace, monospace;
font-size: 1em;
}
input {
color: inherit;
font: inherit;
margin: 0;
}
html input[disabled] {
cursor: default;
}
input {
line-height: normal;
}
input[type="checkbox"] {
box-sizing: border-box;
padding: 0;
}
table {
border-collapse: collapse;
border-spacing: 0;
}
td,
th {
padding: 0;
}
* {
box-sizing: border-box;
}
input {
font: 13px / 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}
a {
color: #4078c0;
text-decoration: none;
}
a:hover,
a:active {
text-decoration: underline;
}
hr {
height: 0;
margin: 15px 0;
overflow: hidden;
background: transparent;
border: 0;
border-bottom: 1px solid #ddd;
}
hr:before {
display: table;
content: "";
}
hr:after {
display: table;
clear: both;
content: "";
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 15px;
margin-bottom: 15px;
line-height: 1.1;
}
h1 {
font-size: 30px;
}
h2 {
font-size: 21px;
}
h3 {
font-size: 16px;
}
h4 {
font-size: 14px;
}
h5 {
font-size: 12px;
}
h6 {
font-size: 11px;
}
blockquote {
margin: 0;
}
ul,
ol {
padding: 0;
margin-top: 0;
margin-bottom: 0;
}
ol ol,
ul ol {
list-style-type: lower-roman;
}
ul ul ol,
ul ol ol,
ol ul ol,
ol ol ol {
list-style-type: lower-alpha;
}
dd {
margin-left: 0;
}
code {
font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
font-size: 12px;
}
pre {
margin-top: 0;
margin-bottom: 0;
font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}
.select::-ms-expand {
opacity: 0;
}
.octicon {
font: normal normal normal 16px/1 octicons-link;
display: inline-block;
text-decoration: none;
text-rendering: auto;
-webkit-font-smoothing: antialiased;
-moz-osx-font-smoothing: grayscale;
-webkit-user-select: none;
-moz-user-select: none;
-ms-user-select: none;
user-select: none;
}
.octicon-link:before {
content: '\f05c';
}
.markdown-body:before {
display: table;
content: "";
}
.markdown-body:after {
display: table;
clear: both;
content: "";
}
.markdown-body>*:first-child {
margin-top: 0 !important;
}
.markdown-body>*:last-child {
margin-bottom: 0 !important;
}
a:not([href]) {
color: inherit;
text-decoration: none;
}
.anchor {
display: inline-block;
padding-right: 2px;
margin-left: -18px;
}
.anchor:focus {
outline: none;
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 1em;
margin-bottom: 16px;
font-weight: bold;
line-height: 1.4;
}
h1 .octicon-link,
h2 .octicon-link,
h3 .octicon-link,
h4 .octicon-link,
h5 .octicon-link,
h6 .octicon-link {
color: #000;
vertical-align: middle;
visibility: hidden;
}
h1:hover .anchor,
h2:hover .anchor,
h3:hover .anchor,
h4:hover .anchor,
h5:hover .anchor,
h6:hover .anchor {
text-decoration: none;
}
h1:hover .anchor .octicon-link,
h2:hover .anchor .octicon-link,
h3:hover .anchor .octicon-link,
h4:hover .anchor .octicon-link,
h5:hover .anchor .octicon-link,
h6:hover .anchor .octicon-link {
visibility: visible;
}
h1 {
padding-bottom: 0.3em;
font-size: 2.25em;
line-height: 1.2;
border-bottom: 1px solid #eee;
}
h1 .anchor {
line-height: 1;
}
h2 {
padding-bottom: 0.3em;
font-size: 1.75em;
line-height: 1.225;
border-bottom: 1px solid #eee;
}
h2 .anchor {
line-height: 1;
}
h3 {
font-size: 1.5em;
line-height: 1.43;
}
h3 .anchor {
line-height: 1.2;
}
h4 {
font-size: 1.25em;
}
h4 .anchor {
line-height: 1.2;
}
h5 {
font-size: 1em;
}
h5 .anchor {
line-height: 1.1;
}
h6 {
font-size: 1em;
color: #777;
}
h6 .anchor {
line-height: 1.1;
}
p,
blockquote,
ul,
ol,
dl,
table,
pre {
margin-top: 0;
margin-bottom: 16px;
}
hr {
height: 4px;
padding: 0;
margin: 16px 0;
background-color: #e7e7e7;
border: 0 none;
}
ul,
ol {
padding-left: 2em;
}
ul ul,
ul ol,
ol ol,
ol ul {
margin-top: 0;
margin-bottom: 0;
}
li>p {
margin-top: 16px;
}
dl {
padding: 0;
}
dl dt {
padding: 0;
margin-top: 16px;
font-size: 1em;
font-style: italic;
font-weight: bold;
}
dl dd {
padding: 0 16px;
margin-bottom: 16px;
}
blockquote {
padding: 0 15px;
color: #777;
border-left: 4px solid #ddd;
}
blockquote>:first-child {
margin-top: 0;
}
blockquote>:last-child {
margin-bottom: 0;
}
table {
display: block;
width: 100%;
overflow: auto;
word-break: normal;
word-break: keep-all;
}
table th {
font-weight: bold;
}
table th,
table td {
padding: 6px 13px;
border: 1px solid #ddd;
}
table tr {
background-color: #fff;
border-top: 1px solid #ccc;
}
table tr:nth-child(2n) {
background-color: #f8f8f8;
}
img {
max-width: 100%;
box-sizing: content-box;
background-color: #fff;
}
code {
padding: 0;
padding-top: 0.2em;
padding-bottom: 0.2em;
margin: 0;
font-size: 85%;
background-color: rgba(0,0,0,0.04);
border-radius: 3px;
}
code:before,
code:after {
letter-spacing: -0.2em;
content: "\00a0";
}
pre>code {
padding: 0;
margin: 0;
font-size: 100%;
word-break: normal;
white-space: pre;
background: transparent;
border: 0;
}
.highlight {
margin-bottom: 16px;
}
.highlight pre,
pre {
padding: 16px;
overflow: auto;
font-size: 85%;
line-height: 1.45;
background-color: #f7f7f7;
border-radius: 3px;
}
.highlight pre {
margin-bottom: 0;
word-break: normal;
}
pre {
word-wrap: normal;
}
pre code {
display: inline;
max-width: initial;
padding: 0;
margin: 0;
overflow: initial;
line-height: inherit;
word-wrap: normal;
background-color: transparent;
border: 0;
}
pre code:before,
pre code:after {
content: normal;
}
kbd {
display: inline-block;
padding: 3px 5px;
font-size: 11px;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.pl-c {
color: #969896;
}
.pl-c1,
.pl-s .pl-v {
color: #0086b3;
}
.pl-e,
.pl-en {
color: #795da3;
}
.pl-s .pl-s1,
.pl-smi {
color: #333;
}
.pl-ent {
color: #63a35c;
}
.pl-k {
color: #a71d5d;
}
.pl-pds,
.pl-s,
.pl-s .pl-pse .pl-s1,
.pl-sr,
.pl-sr .pl-cce,
.pl-sr .pl-sra,
.pl-sr .pl-sre {
color: #183691;
}
.pl-v {
color: #ed6a43;
}
.pl-id {
color: #b52a1d;
}
.pl-ii {
background-color: #b52a1d;
color: #f8f8f8;
}
.pl-sr .pl-cce {
color: #63a35c;
font-weight: bold;
}
.pl-ml {
color: #693a17;
}
.pl-mh,
.pl-mh .pl-en,
.pl-ms {
color: #1d3e81;
font-weight: bold;
}
.pl-mq {
color: #008080;
}
.pl-mi {
color: #333;
font-style: italic;
}
.pl-mb {
color: #333;
font-weight: bold;
}
.pl-md {
background-color: #ffecec;
color: #bd2c00;
}
.pl-mi1 {
background-color: #eaffea;
color: #55a532;
}
.pl-mdr {
color: #795da3;
font-weight: bold;
}
.pl-mo {
color: #1d3e81;
}
kbd {
display: inline-block;
padding: 3px 5px;
font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.task-list-item {
list-style-type: none;
}
.task-list-item+.task-list-item {
margin-top: 3px;
}
.task-list-item input {
margin: 0 0.35em 0.25em -1.6em;
vertical-align: middle;
}
:checked+.radio-label {
z-index: 1;
position: relative;
border-color: #4078c0;
}
.sourceLine {
display: inline-block;
}
code .kw { color: #000000; }
code .dt { color: #ed6a43; }
code .dv { color: #009999; }
code .bn { color: #009999; }
code .fl { color: #009999; }
code .ch { color: #009999; }
code .st { color: #183691; }
code .co { color: #969896; }
code .ot { color: #0086b3; }
code .al { color: #a61717; }
code .fu { color: #63a35c; }
code .er { color: #a61717; background-color: #e3d2d2; }
code .wa { color: #000000; }
code .cn { color: #008080; }
code .sc { color: #008080; }
code .vs { color: #183691; }
code .ss { color: #183691; }
code .im { color: #000000; }
code .va {color: #008080; }
code .cf { color: #000000; }
code .op { color: #000000; }
code .bu { color: #000000; }
code .ex { color: #000000; }
code .pp { color: #999999; }
code .at { color: #008080; }
code .do { color: #969896; }
code .an { color: #008080; }
code .cv { color: #008080; }
code .in { color: #008080; }
</style>
<style>
body {
box-sizing: border-box;
min-width: 200px;
max-width: 980px;
margin: 0 auto;
padding: 45px;
padding-top: 0px;
}
</style>


</head>

<body>

<h1 id="genular-an-integrated-knowledgebase-platform-for-revealing-cell-specific-genetic-signatures-and-functional-insights">Genular:
An Integrated Knowledgebase Platform for Revealing Cell-Specific Genetic
Signatures and Functional Insights</h1>
<p>Ivan Tomic <a href="mailto:info@ivantomic.com">info@ivantomic.com</a></p>
<!-- README.md is generated from README.Rmd. Please edit that file -->

<p>The <code>genular</code> package provides a comprehensive toolkit for
interacting with the <a href="https://genular.atomic-lab.org">Genular
Database API</a>, enabling efficient retrieval, analysis, and
exploration of genomic and single-cell expression data directly within
the R environment.</p>
<h2 id="installation">Installation</h2>
<p>You can install the released version of <code>genular</code> from <a href="https://CRAN.R-project.org">CRAN</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;genular&quot;</span>)</span></code></pre></div>
<p>Or you can install <code>genular</code> directly from GitHub with use
of following commands:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;atomiclaboratory/genular-database&quot;</span>, <span class="at">subdir =</span> <span class="st">&#39;libraries/genular-api/R-package&#39;</span>)</span></code></pre></div>
<h2 id="usage-examples">Usage Examples</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Check if plyr is installed and install it if necessary</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;plyr&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;plyr&quot;</span>)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>}</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;dplyr&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;dplyr&quot;</span>)</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>}</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co"># Load the genular r-package</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="fu">library</span>(genular)</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="co"># Your personal API Key </span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="co"># (request here or use your own installation: https://genular.atomic-lab.org/contact)</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>API_KEY <span class="ot">&lt;-</span> <span class="st">&quot;3147dda5fa023c9763d38bddb472dd28&quot;</span></span></code></pre></div>
<h2 id="example-1-suggesting-cell-matches-based-on-query-values">Example
1: Suggesting Cell Matches Based on Query Values</h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># This example demonstrates how to suggest relevant cell types based on provided keywords. We use the `cells_suggest()` function to find cell types related to specific terms like &quot;endothelial cell&quot; and &quot;T cell&quot;.</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co"># Define the query values for cell type suggestions</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>queryValues <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;endothelial cell&quot;</span>, <span class="st">&quot;T cell&quot;</span>)</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co"># Use the cells_suggest function to get relevant cell type matches</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>search_example_1 <span class="ot">&lt;-</span> <span class="fu">cells_suggest</span>(queryValues,</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>                                        <span class="at">options =</span> <span class="fu">list</span>(<span class="at">api_key =</span> API_KEY, <span class="at">timeout =</span> <span class="dv">5000</span>)</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>                                      )</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co"># Display the suggested cell types</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co"># print(head(search_example_1))</span></span></code></pre></div>
<h2 id="example-2-searching-cell-information-based-on-query-conditions">Example
2: Searching Cell Information Based on Query Conditions</h2>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># In this example, we demonstrate how to search for specific cell information based on conditions applied to multiple cell IDs. The `cells_search()` function is used to retrieve data about cells that meet the specified criteria.</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co"># Define query conditions for different cell IDs</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>queryValues <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>  <span class="st">&quot;CL0002350&quot;</span> <span class="ot">=</span> <span class="st">&quot;&gt;= 565&quot;</span>,</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>  <span class="st">&quot;CL2000041&quot;</span> <span class="ot">=</span> <span class="st">&quot;&gt;= 323&quot;</span>,</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>  <span class="st">&quot;CL0000129&quot;</span> <span class="ot">=</span> <span class="st">&quot;&gt;= 1821&quot;</span>,</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>  <span class="st">&quot;CL0000453&quot;</span> <span class="ot">=</span> <span class="st">&quot;&gt;= 299&quot;</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>)</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co"># Define the fields to filter in the search results</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>fieldsFilter <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>  <span class="st">&quot;geneID&quot;</span>, </span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>  <span class="st">&quot;symbol&quot;</span>, </span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a>  <span class="st">&quot;crossReference.enseGeneID&quot;</span>, </span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a>  <span class="st">&quot;singleCellExpressions.effectSizes.i&quot;</span>, </span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a>  <span class="st">&quot;singleCellExpressions.effectSizes.s&quot;</span>, </span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>  <span class="st">&quot;singleCellExpressions.effectSizes.c&quot;</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a>)</span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a><span class="co"># Perform the cell search</span></span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a>search_example_2 <span class="ot">&lt;-</span> <span class="fu">cells_search</span>(</span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a>  queryValues, </span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a>  fieldsFilter,</span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a>  <span class="at">page =</span> <span class="dv">1</span>,</span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a>  <span class="at">limit =</span> <span class="dv">10</span>,</span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a>  <span class="at">searchType =</span> <span class="st">&quot;or&quot;</span>,</span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a>  <span class="at">orderBy =</span> <span class="st">&quot;geneID&quot;</span>,</span>
<span id="cb5-29"><a href="#cb5-29" tabindex="-1"></a>  <span class="at">sortDirection =</span> <span class="st">&quot;asc&quot;</span>,</span>
<span id="cb5-30"><a href="#cb5-30" tabindex="-1"></a>  <span class="at">responseType =</span> <span class="st">&quot;json&quot;</span>,</span>
<span id="cb5-31"><a href="#cb5-31" tabindex="-1"></a>  <span class="at">organismType =</span> <span class="fu">list</span>(<span class="dv">9606</span>),</span>
<span id="cb5-32"><a href="#cb5-32" tabindex="-1"></a>  <span class="at">options =</span> <span class="fu">list</span>(<span class="at">api_key =</span> <span class="st">&quot;3147dda5fa023c9763d38bddb472dd28&quot;</span>, <span class="at">timeout =</span> <span class="dv">10000</span>)</span>
<span id="cb5-33"><a href="#cb5-33" tabindex="-1"></a>)</span>
<span id="cb5-34"><a href="#cb5-34" tabindex="-1"></a><span class="co"># Display the results</span></span>
<span id="cb5-35"><a href="#cb5-35" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">head</span>(<span class="fu">sapply</span>(search_example_2<span class="sc">$</span>content<span class="sc">$</span>results, <span class="cf">function</span>(x) x<span class="sc">$</span>symbol)))</span>
<span id="cb5-36"><a href="#cb5-36" tabindex="-1"></a><span class="co">#&gt; [1] &quot;A2M&quot;    &quot;ABR&quot;    &quot;ACAA1&quot;  &quot;ACADVL&quot; &quot;ASIC2&quot;  &quot;ACP1&quot;</span></span></code></pre></div>
<h2 id="example-3-searching-for-gene-information">Example 3: Searching
for Gene Information</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># This example demonstrates how to search for specific gene information using gene IDs. We use the `gene_search()` function to query gene-related details such as gene symbols and Ensembl gene IDs.</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co"># Define the fields to search for and query gene IDs</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>queryFields <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="st">&quot;geneID&quot;</span>))</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>queryValues <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">56</span>, <span class="dv">70</span>)  <span class="co"># Specify the gene IDs to search for</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>searchType <span class="ot">&lt;-</span> <span class="st">&quot;or&quot;</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>fieldsFilter <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;geneID&quot;</span>, <span class="st">&quot;symbol&quot;</span>, <span class="st">&quot;crossReference.enseGeneID&quot;</span>)</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>organismType <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">9606</span>)</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co"># Perform the gene search</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>search_example_3 <span class="ot">&lt;-</span> <span class="fu">gene_search</span>(queryFields, queryValues, fieldsFilter, <span class="at">searchType =</span> searchType,</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>                                    <span class="at">organismType =</span> organismType, <span class="at">page =</span> <span class="dv">1</span>, <span class="at">limit =</span> <span class="dv">10</span>,</span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>                                    <span class="at">options =</span> <span class="fu">list</span>(<span class="at">api_key =</span> API_KEY, <span class="at">timeout =</span> <span class="dv">1000</span>))</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="co"># Display the results</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">head</span>(<span class="fu">sapply</span>(search_example_3<span class="sc">$</span>content<span class="sc">$</span>results, <span class="cf">function</span>(x) x<span class="sc">$</span>symbol)))</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="co">#&gt; [1] &quot;A1BG&quot;  &quot;ACRV1&quot; &quot;ACTC1&quot;</span></span></code></pre></div>
<h2 id="example-4-suggesting-pathway-matches-based-on-query-values">Example
4: Suggesting Pathway Matches Based on Query Values</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># In this example, we demonstrate how to use keywords to suggest relevant biological pathways. We use the `pathways_suggest()` function to find pathways related to specific biological processes or activities, such as &quot;apoptosis&quot; and &quot;signal transduction&quot;.</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co"># Define the query values for pathway suggestion</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>queryValues <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;apoptosis&quot;</span>, <span class="st">&quot;signal transduction&quot;</span>)</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="co"># Use the pathways_suggest function to get relevant pathway matches</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>search_example_4 <span class="ot">&lt;-</span> <span class="fu">pathways_suggest</span>(queryValues)</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co"># Display the suggested pathways</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">head</span>(<span class="fu">sapply</span>(search_example_4<span class="sc">$</span>results, <span class="cf">function</span>(x) x<span class="sc">$</span>term)))</span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Apoptosis&quot;                           &quot;Signal Transduction&quot;                </span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="co">#&gt; [3] &quot;Signal transduction by L1&quot;           &quot;Apoptosis induced DNA fragmentation&quot;</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="co">#&gt; [5] &quot;Signaling by NODAL&quot;                  &quot;Signaling by FGFR in disease&quot;</span></span></code></pre></div>
<h2 id="example-5-gene-expression-feature-engineering-pathway-level-features--pathwaygenescore">Example
5: Gene Expression Feature Engineering (Pathway-Level Features /
PathwayGeneScore)</h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># This example demonstrates how to perform feature engineering on gene expression data using external gene-related information. The goal is to convert raw gene expression data into pathway-level features for further analysis.</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="co"># Define gene expression data for a set of genes across multiple samples</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co"># Demo data HIPC - Expression data with meta</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>remote_data <span class="ot">&lt;-</span> <span class="st">&quot;https://genular.ams3.cdn.digitaloceanspaces.com/database-export/hipc_expression_data_with_meta.csv&quot;</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>input_data <span class="ot">&lt;-</span> <span class="fu">read.csv2</span>(remote_data, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;,&quot;</span>, <span class="at">dec =</span> <span class="st">&quot;.&quot;</span>, <span class="at">quote =</span> <span class="st">&quot;</span><span class="sc">\&quot;</span><span class="st">&quot;</span>, <span class="at">fill =</span> <span class="cn">TRUE</span>, <span class="at">comment.char =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="fu">library</span>(stats)</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co"># START DATA PREPROCESSING</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co"># This is just an example of how to preprocess the data before using the genular API</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co"># If you have raw gene expression data, probbably you will need to preprocess it before using the Genular API (eg. DESeq2)</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> input_data[, <span class="dv">1</span><span class="sc">:</span><span class="dv">36</span>]  <span class="co"># Extract the first 36 columns as metadata</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a>input_data_counts <span class="ot">&lt;-</span> input_data[, <span class="dv">37</span><span class="sc">:</span><span class="fu">ncol</span>(input_data)]  <span class="co"># Extract gene expression data</span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a>scaled_data <span class="ot">&lt;-</span> <span class="fu">scale</span>(input_data_counts)</span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="co"># Perform PCA on the scaled data</span></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a>pca_results <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(scaled_data, <span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale. =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="co"># Get the explained variance proportions for each principal component</span></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a>explained_variance <span class="ot">&lt;-</span> <span class="fu">summary</span>(pca_results)<span class="sc">$</span>importance[<span class="dv">2</span>,]</span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a><span class="co"># Find the number of principal components explaining 90% of the variance</span></span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a>cumulative_variance <span class="ot">&lt;-</span> <span class="fu">cumsum</span>(explained_variance)</span>
<span id="cb8-24"><a href="#cb8-24" tabindex="-1"></a>num_components <span class="ot">&lt;-</span> <span class="fu">which</span>(cumulative_variance <span class="sc">&gt;=</span> <span class="fl">0.90</span>)[<span class="dv">1</span>]</span>
<span id="cb8-25"><a href="#cb8-25" tabindex="-1"></a><span class="co"># Get the PCA loadings (rotation matrix)</span></span>
<span id="cb8-26"><a href="#cb8-26" tabindex="-1"></a>loadings <span class="ot">&lt;-</span> pca_results<span class="sc">$</span>rotation[, <span class="dv">1</span><span class="sc">:</span>num_components]</span>
<span id="cb8-27"><a href="#cb8-27" tabindex="-1"></a><span class="co"># Calculate the variance contribution of each variable by squaring the loadings</span></span>
<span id="cb8-28"><a href="#cb8-28" tabindex="-1"></a><span class="co"># and multiplying by the variance explained by the respective component</span></span>
<span id="cb8-29"><a href="#cb8-29" tabindex="-1"></a>variance_contributions <span class="ot">&lt;-</span> <span class="fu">rowSums</span>((loadings<span class="sc">^</span><span class="dv">2</span>) <span class="sc">*</span> explained_variance[<span class="dv">1</span><span class="sc">:</span>num_components])</span>
<span id="cb8-30"><a href="#cb8-30" tabindex="-1"></a><span class="co"># Rank variables by their contribution to the variance</span></span>
<span id="cb8-31"><a href="#cb8-31" tabindex="-1"></a>ranked_variables <span class="ot">&lt;-</span> <span class="fu">sort</span>(variance_contributions, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-32"><a href="#cb8-32" tabindex="-1"></a><span class="co"># Determine how many variables to retain (e.g., 90% of the cumulative variance contribution)</span></span>
<span id="cb8-33"><a href="#cb8-33" tabindex="-1"></a>cumulative_var_contrib <span class="ot">&lt;-</span> <span class="fu">cumsum</span>(ranked_variables) <span class="sc">/</span> <span class="fu">sum</span>(ranked_variables)</span>
<span id="cb8-34"><a href="#cb8-34" tabindex="-1"></a><span class="co"># Find the number of variables that cumulatively contribute to 0.75% of the variance explained by these variables</span></span>
<span id="cb8-35"><a href="#cb8-35" tabindex="-1"></a>num_important_variables <span class="ot">&lt;-</span> <span class="fu">which</span>(cumulative_var_contrib <span class="sc">&gt;=</span> <span class="fl">0.75</span>)[<span class="dv">1</span>]</span>
<span id="cb8-36"><a href="#cb8-36" tabindex="-1"></a><span class="co"># Select the top contributing variables</span></span>
<span id="cb8-37"><a href="#cb8-37" tabindex="-1"></a>important_variables <span class="ot">&lt;-</span> <span class="fu">names</span>(ranked_variables[<span class="dv">1</span><span class="sc">:</span>num_important_variables])</span>
<span id="cb8-38"><a href="#cb8-38" tabindex="-1"></a><span class="co"># Subset input_data_counts to include only the important variables</span></span>
<span id="cb8-39"><a href="#cb8-39" tabindex="-1"></a>input_data_filtered <span class="ot">&lt;-</span> input_data_counts[, important_variables]</span>
<span id="cb8-40"><a href="#cb8-40" tabindex="-1"></a></span>
<span id="cb8-41"><a href="#cb8-41" tabindex="-1"></a><span class="co"># </span><span class="re">END</span><span class="co"> DATA PREPROCESSING</span></span>
<span id="cb8-42"><a href="#cb8-42" tabindex="-1"></a></span>
<span id="cb8-43"><a href="#cb8-43" tabindex="-1"></a><span class="do">### Fetch gene-related data from genular database</span></span>
<span id="cb8-44"><a href="#cb8-44" tabindex="-1"></a><span class="co"># Using `fetch_all_gene_search_results()`, we query gene information for the given symbols in `input_data`</span></span>
<span id="cb8-45"><a href="#cb8-45" tabindex="-1"></a>search_example_5 <span class="ot">&lt;-</span> <span class="fu">fetch_all_gene_search_results</span>(</span>
<span id="cb8-46"><a href="#cb8-46" tabindex="-1"></a>  <span class="at">queryFields =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="st">&quot;symbol&quot;</span>)),  <span class="co"># Specify that the query is based on gene symbols</span></span>
<span id="cb8-47"><a href="#cb8-47" tabindex="-1"></a>  <span class="at">queryValues =</span> <span class="fu">colnames</span>(input_data_filtered),  <span class="co"># Gene symbols extracted from the column names of &#39;input_data&#39;</span></span>
<span id="cb8-48"><a href="#cb8-48" tabindex="-1"></a>  <span class="at">fieldsFilter =</span> <span class="fu">c</span>(<span class="st">&quot;geneID&quot;</span>, <span class="st">&quot;symbol&quot;</span>, <span class="st">&quot;crossReference.enseGeneID&quot;</span>, </span>
<span id="cb8-49"><a href="#cb8-49" tabindex="-1"></a>                   <span class="st">&quot;mRNAExpressions.proteinAtlas.c&quot;</span>, </span>
<span id="cb8-50"><a href="#cb8-50" tabindex="-1"></a>                   <span class="st">&quot;ontology.id&quot;</span>, <span class="st">&quot;ontology.term&quot;</span>, <span class="st">&quot;ontology.cat&quot;</span>),  <span class="co"># Fields to be fetched</span></span>
<span id="cb8-51"><a href="#cb8-51" tabindex="-1"></a>  <span class="at">searchType =</span> <span class="st">&quot;or&quot;</span>,  <span class="co"># Search type specifying how to combine query fields</span></span>
<span id="cb8-52"><a href="#cb8-52" tabindex="-1"></a>  <span class="at">orderBy =</span> <span class="st">&quot;geneID&quot;</span>,  <span class="co"># Order the results by gene ID</span></span>
<span id="cb8-53"><a href="#cb8-53" tabindex="-1"></a>  <span class="at">sortDirection =</span> <span class="st">&quot;asc&quot;</span>,  <span class="co"># Sort direction: ascending</span></span>
<span id="cb8-54"><a href="#cb8-54" tabindex="-1"></a>  <span class="at">responseType =</span> <span class="st">&quot;json&quot;</span>,  <span class="co"># Format of the response</span></span>
<span id="cb8-55"><a href="#cb8-55" tabindex="-1"></a>  <span class="at">matchType =</span> <span class="st">&quot;exact&quot;</span>,  <span class="co"># Type of matching to use for the query</span></span>
<span id="cb8-56"><a href="#cb8-56" tabindex="-1"></a>  <span class="at">organismType =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="dv">9606</span>)),  <span class="co"># Specify the organism type (e.g., Homo sapiens)</span></span>
<span id="cb8-57"><a href="#cb8-57" tabindex="-1"></a>  <span class="at">ontologyCategories =</span> <span class="fu">list</span>(),  <span class="co"># Specify ontology categories if needed</span></span>
<span id="cb8-58"><a href="#cb8-58" tabindex="-1"></a>  <span class="at">limit =</span> <span class="dv">100</span>,  <span class="co"># Limit the number of results returned</span></span>
<span id="cb8-59"><a href="#cb8-59" tabindex="-1"></a>  <span class="at">options =</span> <span class="fu">list</span>(<span class="at">api_key =</span> API_KEY, <span class="at">timeout =</span> <span class="dv">1000</span>)  <span class="co"># API key and timeout for the request</span></span>
<span id="cb8-60"><a href="#cb8-60" tabindex="-1"></a>)</span>
<span id="cb8-61"><a href="#cb8-61" tabindex="-1"></a></span>
<span id="cb8-62"><a href="#cb8-62" tabindex="-1"></a>all_gene_symbols <span class="ot">&lt;-</span> <span class="fu">sapply</span>(search_example_5, <span class="cf">function</span>(x) x<span class="sc">$</span>symbol)</span>
<span id="cb8-63"><a href="#cb8-63" tabindex="-1"></a>gene_not_found <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="fu">colnames</span>(input_data_filtered), all_gene_symbols)</span>
<span id="cb8-64"><a href="#cb8-64" tabindex="-1"></a>input_data_filtered <span class="ot">&lt;-</span> input_data_filtered[, <span class="fu">setdiff</span>(<span class="fu">colnames</span>(input_data_filtered), gene_not_found)]</span>
<span id="cb8-65"><a href="#cb8-65" tabindex="-1"></a></span>
<span id="cb8-66"><a href="#cb8-66" tabindex="-1"></a><span class="do">## Transform and restructure the fetched gene-related data</span></span>
<span id="cb8-67"><a href="#cb8-67" tabindex="-1"></a><span class="co"># The `extract_data()` function helps map various gene attributes to a more usable format</span></span>
<span id="cb8-68"><a href="#cb8-68" tabindex="-1"></a>data_transposed <span class="ot">&lt;-</span> <span class="fu">extract_data</span>(search_example_5, <span class="fu">list</span>(</span>
<span id="cb8-69"><a href="#cb8-69" tabindex="-1"></a>    <span class="st">&quot;geneID&quot;</span> <span class="ot">=</span> <span class="st">&quot;mappedGeneID&quot;</span>,  <span class="co"># Mapping gene IDs</span></span>
<span id="cb8-70"><a href="#cb8-70" tabindex="-1"></a>    <span class="st">&quot;symbol&quot;</span> <span class="ot">=</span> <span class="st">&quot;mappedSymbol&quot;</span>,  <span class="co"># Mapping gene symbols</span></span>
<span id="cb8-71"><a href="#cb8-71" tabindex="-1"></a>    <span class="st">&quot;crossReference$enseGeneID&quot;</span> <span class="ot">=</span> <span class="st">&quot;mappedEnseGeneID&quot;</span>,  <span class="co"># Mapping Ensembl gene IDs</span></span>
<span id="cb8-72"><a href="#cb8-72" tabindex="-1"></a>    <span class="st">&quot;mRNAExpressions$proteinAtlas&quot;</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="st">&quot;c&quot;</span> <span class="ot">=</span> <span class="st">&quot;mappedC&quot;</span>)),  <span class="co"># Mapping mRNA expression</span></span>
<span id="cb8-73"><a href="#cb8-73" tabindex="-1"></a>    <span class="st">&quot;ontology&quot;</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="st">&quot;id&quot;</span> <span class="ot">=</span> <span class="st">&quot;mappedId&quot;</span>, <span class="st">&quot;term&quot;</span> <span class="ot">=</span> <span class="st">&quot;mappedTerm&quot;</span>, <span class="st">&quot;cat&quot;</span> <span class="ot">=</span> <span class="st">&quot;mappedCat&quot;</span>))  <span class="co"># Mapping ontology data</span></span>
<span id="cb8-74"><a href="#cb8-74" tabindex="-1"></a>))</span>
<span id="cb8-75"><a href="#cb8-75" tabindex="-1"></a></span>
<span id="cb8-76"><a href="#cb8-76" tabindex="-1"></a><span class="co"># Remove rows with missing ontology information</span></span>
<span id="cb8-77"><a href="#cb8-77" tabindex="-1"></a>data_transposed <span class="ot">&lt;-</span> data_transposed[<span class="sc">!</span><span class="fu">is.na</span>(data_transposed<span class="sc">$</span>mappedId), ]</span>
<span id="cb8-78"><a href="#cb8-78" tabindex="-1"></a><span class="co"># View the first few rows of the transposed data</span></span>
<span id="cb8-79"><a href="#cb8-79" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">head</span>(data_transposed[<span class="fu">order</span>(data_transposed<span class="sc">$</span>mappedId), ]))</span>
<span id="cb8-80"><a href="#cb8-80" tabindex="-1"></a></span>
<span id="cb8-81"><a href="#cb8-81" tabindex="-1"></a><span class="do">## Convert gene expression data to pathway-level features / this can take a while</span></span>
<span id="cb8-82"><a href="#cb8-82" tabindex="-1"></a><span class="co"># Using the `convert_gene_expression_to_pathway_features()` function, we generate PathwayGeneScore and pathway-level features</span></span>
<span id="cb8-83"><a href="#cb8-83" tabindex="-1"></a>final_data <span class="ot">&lt;-</span> <span class="fu">convert_gene_expression_to_pathway_features</span>(input_data_filtered, data_transposed, T)</span>
<span id="cb8-84"><a href="#cb8-84" tabindex="-1"></a></span>
<span id="cb8-85"><a href="#cb8-85" tabindex="-1"></a><span class="co"># (Optional) Remove NA columns from the final data</span></span>
<span id="cb8-86"><a href="#cb8-86" tabindex="-1"></a>final_data <span class="ot">&lt;-</span> final_data[, <span class="fu">colSums</span>(<span class="sc">!</span><span class="fu">is.na</span>(final_data)) <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb8-87"><a href="#cb8-87" tabindex="-1"></a></span>
<span id="cb8-88"><a href="#cb8-88" tabindex="-1"></a><span class="co"># View the final data</span></span>
<span id="cb8-89"><a href="#cb8-89" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">head</span>(final_data))</span>
<span id="cb8-90"><a href="#cb8-90" tabindex="-1"></a><span class="co"># Continue with downstream analysis using the pathway-level features</span></span></code></pre></div>
<h2 id="example-6-get-all-genes-for-adaptive-immune-system-and-their-corresponding-cell-marker-scores">Example
6: Get all Genes for “Adaptive Immune System” and Their Corresponding
Cell Marker Scores</h2>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># This example demonstrates how to retrieve genes associated with the &quot;Adaptive Immune System&quot; and process their corresponding cell marker scores for further analysis. We use `fetch_all_gene_search_results()` to obtain gene information, followed by data filtering and visualization steps.</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="co"># Fetch genes related to &quot;Adaptive Immune System&quot;</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>search_example_6 <span class="ot">&lt;-</span> <span class="fu">fetch_all_gene_search_results</span>(</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>  <span class="at">queryFields =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="st">&quot;ontology.term&quot;</span>)),  </span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>  <span class="at">queryValues =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="st">&quot;Adaptive Immune System&quot;</span>)),  </span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>  <span class="at">fieldsFilter =</span> <span class="fu">c</span>(<span class="st">&quot;geneID&quot;</span>, <span class="st">&quot;symbol&quot;</span>, <span class="st">&quot;crossReference.enseGeneID&quot;</span>, </span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>                   <span class="st">&quot;singleCellExpressions.effectSizes.i&quot;</span>, </span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>                   <span class="st">&quot;singleCellExpressions.effectSizes.c&quot;</span>, </span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>                   <span class="st">&quot;singleCellExpressions.effectSizes.s&quot;</span>,</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>                   <span class="st">&quot;singleCellExpressions.effectSizes.foldChange&quot;</span>),  </span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>  <span class="at">searchType =</span> <span class="st">&quot;or&quot;</span>,  </span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>  <span class="at">orderBy =</span> <span class="st">&quot;geneID&quot;</span>,  </span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a>  <span class="at">sortDirection =</span> <span class="st">&quot;asc&quot;</span>,  </span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a>  <span class="at">responseType =</span> <span class="st">&quot;json&quot;</span>,  </span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a>  <span class="at">matchType =</span> <span class="st">&quot;exact&quot;</span>,  </span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a>  <span class="at">organismType =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="dv">9606</span>)),  </span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a>  <span class="at">limit =</span> <span class="dv">100</span>,  </span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a>  <span class="at">options =</span> <span class="fu">list</span>(<span class="at">api_key =</span> API_KEY, <span class="at">timeout =</span> <span class="dv">1000</span>)</span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a>)</span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a></span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a></span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a><span class="fu">library</span>(purrr)</span>
<span id="cb9-26"><a href="#cb9-26" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb9-27"><a href="#cb9-27" tabindex="-1"></a></span>
<span id="cb9-28"><a href="#cb9-28" tabindex="-1"></a>filter_effect_sizes_top_percentile <span class="ot">&lt;-</span> <span class="cf">function</span>(gene, <span class="at">percentile =</span> <span class="dv">90</span>) {</span>
<span id="cb9-29"><a href="#cb9-29" tabindex="-1"></a>  <span class="co"># Extract all foldChange values for the current gene</span></span>
<span id="cb9-30"><a href="#cb9-30" tabindex="-1"></a>  fold_changes <span class="ot">&lt;-</span> <span class="fu">map_dbl</span>(gene<span class="sc">$</span>singleCellExpressions<span class="sc">$</span>effectSizes, <span class="sc">~</span> .x<span class="sc">$</span>foldChange)</span>
<span id="cb9-31"><a href="#cb9-31" tabindex="-1"></a>  </span>
<span id="cb9-32"><a href="#cb9-32" tabindex="-1"></a>  <span class="co"># Calculate the specified percentile (default is 90th percentile)</span></span>
<span id="cb9-33"><a href="#cb9-33" tabindex="-1"></a>  cutoff <span class="ot">&lt;-</span> <span class="fu">quantile</span>(fold_changes, <span class="at">probs =</span> percentile <span class="sc">/</span> <span class="dv">100</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-34"><a href="#cb9-34" tabindex="-1"></a>  </span>
<span id="cb9-35"><a href="#cb9-35" tabindex="-1"></a>  <span class="co"># Filter effectSizes to retain only those with foldChange &gt;= cutoff</span></span>
<span id="cb9-36"><a href="#cb9-36" tabindex="-1"></a>  filtered_effect_sizes <span class="ot">&lt;-</span> gene<span class="sc">$</span>singleCellExpressions<span class="sc">$</span>effectSizes <span class="sc">%&gt;%</span></span>
<span id="cb9-37"><a href="#cb9-37" tabindex="-1"></a>    <span class="fu">keep</span>(<span class="sc">~</span> .x<span class="sc">$</span>foldChange <span class="sc">&gt;=</span> cutoff)</span>
<span id="cb9-38"><a href="#cb9-38" tabindex="-1"></a>  </span>
<span id="cb9-39"><a href="#cb9-39" tabindex="-1"></a>  <span class="co"># Update the gene&#39;s effectSizes with the filtered list</span></span>
<span id="cb9-40"><a href="#cb9-40" tabindex="-1"></a>  gene<span class="sc">$</span>singleCellExpressions<span class="sc">$</span>effectSizes <span class="ot">&lt;-</span> filtered_effect_sizes</span>
<span id="cb9-41"><a href="#cb9-41" tabindex="-1"></a>  </span>
<span id="cb9-42"><a href="#cb9-42" tabindex="-1"></a>  <span class="fu">return</span>(gene)</span>
<span id="cb9-43"><a href="#cb9-43" tabindex="-1"></a>}</span>
<span id="cb9-44"><a href="#cb9-44" tabindex="-1"></a></span>
<span id="cb9-45"><a href="#cb9-45" tabindex="-1"></a>filter_effect_sizes_by_lineage <span class="ot">&lt;-</span> <span class="cf">function</span>(gene, <span class="at">lineage =</span> <span class="st">&quot;root&quot;</span>) {</span>
<span id="cb9-46"><a href="#cb9-46" tabindex="-1"></a>  <span class="co"># Filter effectSizes to keep only those with cell_lineage == &#39;root&#39;</span></span>
<span id="cb9-47"><a href="#cb9-47" tabindex="-1"></a>  filtered_effect_sizes <span class="ot">&lt;-</span> gene<span class="sc">$</span>singleCellExpressions<span class="sc">$</span>effectSizes <span class="sc">%&gt;%</span></span>
<span id="cb9-48"><a href="#cb9-48" tabindex="-1"></a>    <span class="fu">keep</span>(<span class="sc">~</span> .x<span class="sc">$</span>cell_lineage <span class="sc">==</span> lineage)</span>
<span id="cb9-49"><a href="#cb9-49" tabindex="-1"></a>  </span>
<span id="cb9-50"><a href="#cb9-50" tabindex="-1"></a>  <span class="co"># Update the gene&#39;s effectSizes with the filtered list</span></span>
<span id="cb9-51"><a href="#cb9-51" tabindex="-1"></a>  gene<span class="sc">$</span>singleCellExpressions<span class="sc">$</span>effectSizes <span class="ot">&lt;-</span> filtered_effect_sizes</span>
<span id="cb9-52"><a href="#cb9-52" tabindex="-1"></a>  </span>
<span id="cb9-53"><a href="#cb9-53" tabindex="-1"></a>  <span class="fu">return</span>(gene)</span>
<span id="cb9-54"><a href="#cb9-54" tabindex="-1"></a>}</span>
<span id="cb9-55"><a href="#cb9-55" tabindex="-1"></a></span>
<span id="cb9-56"><a href="#cb9-56" tabindex="-1"></a>search_example_6_filtered <span class="ot">&lt;-</span> search_example_6 <span class="sc">%&gt;%</span></span>
<span id="cb9-57"><a href="#cb9-57" tabindex="-1"></a>  <span class="fu">map</span>(<span class="sc">~</span> <span class="fu">filter_effect_sizes_top_percentile</span>(.x, <span class="at">percentile =</span> <span class="dv">95</span>))</span>
<span id="cb9-58"><a href="#cb9-58" tabindex="-1"></a></span>
<span id="cb9-59"><a href="#cb9-59" tabindex="-1"></a>search_example_6_filtered <span class="ot">&lt;-</span> search_example_6_filtered <span class="sc">%&gt;%</span></span>
<span id="cb9-60"><a href="#cb9-60" tabindex="-1"></a>  <span class="fu">map</span>(<span class="sc">~</span> <span class="fu">filter_effect_sizes_by_lineage</span>(.x, <span class="at">lineage =</span> <span class="st">&quot;root&quot;</span>))</span>
<span id="cb9-61"><a href="#cb9-61" tabindex="-1"></a></span>
<span id="cb9-62"><a href="#cb9-62" tabindex="-1"></a><span class="co"># Remove genes with no remaining effectSizes</span></span>
<span id="cb9-63"><a href="#cb9-63" tabindex="-1"></a>search_example_6_filtered <span class="ot">&lt;-</span> search_example_6_filtered <span class="sc">%&gt;%</span></span>
<span id="cb9-64"><a href="#cb9-64" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">keep</span>(<span class="sc">~</span> <span class="fu">length</span>(.x<span class="sc">$</span>singleCellExpressions<span class="sc">$</span>effectSizes) <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb9-65"><a href="#cb9-65" tabindex="-1"></a></span>
<span id="cb9-66"><a href="#cb9-66" tabindex="-1"></a><span class="co"># Sort genes by the number of cells in descending order</span></span>
<span id="cb9-67"><a href="#cb9-67" tabindex="-1"></a>num_cells_per_gene <span class="ot">&lt;-</span> <span class="fu">sapply</span>(search_example_6_filtered, <span class="cf">function</span>(gene) {</span>
<span id="cb9-68"><a href="#cb9-68" tabindex="-1"></a>  <span class="fu">length</span>(gene<span class="sc">$</span>singleCellExpressions<span class="sc">$</span>effectSizes)</span>
<span id="cb9-69"><a href="#cb9-69" tabindex="-1"></a>})</span>
<span id="cb9-70"><a href="#cb9-70" tabindex="-1"></a>ordered_indices <span class="ot">&lt;-</span> <span class="fu">order</span>(num_cells_per_gene, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-71"><a href="#cb9-71" tabindex="-1"></a>search_example_6_filtered_sorted <span class="ot">&lt;-</span> search_example_6_filtered[ordered_indices]</span>
<span id="cb9-72"><a href="#cb9-72" tabindex="-1"></a></span>
<span id="cb9-73"><a href="#cb9-73" tabindex="-1"></a><span class="co"># Convert the sorted and filtered list to a tidy data frame for visualization</span></span>
<span id="cb9-74"><a href="#cb9-74" tabindex="-1"></a>df_final <span class="ot">&lt;-</span> search_example_6_filtered_sorted <span class="sc">%&gt;%</span></span>
<span id="cb9-75"><a href="#cb9-75" tabindex="-1"></a>  <span class="fu">map_dfr</span>(<span class="cf">function</span>(gene) {</span>
<span id="cb9-76"><a href="#cb9-76" tabindex="-1"></a>    gene_info <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb9-77"><a href="#cb9-77" tabindex="-1"></a>      <span class="at">geneID =</span> gene<span class="sc">$</span>geneID,</span>
<span id="cb9-78"><a href="#cb9-78" tabindex="-1"></a>      <span class="at">symbol =</span> gene<span class="sc">$</span>symbol,</span>
<span id="cb9-79"><a href="#cb9-79" tabindex="-1"></a>      <span class="at">ensemblID =</span> gene<span class="sc">$</span>crossReference<span class="sc">$</span>enseGeneID</span>
<span id="cb9-80"><a href="#cb9-80" tabindex="-1"></a>    )</span>
<span id="cb9-81"><a href="#cb9-81" tabindex="-1"></a>    effect_sizes <span class="ot">&lt;-</span> gene<span class="sc">$</span>singleCellExpressions<span class="sc">$</span>effectSizes <span class="sc">%&gt;%</span></span>
<span id="cb9-82"><a href="#cb9-82" tabindex="-1"></a>      <span class="fu">map_dfr</span>(<span class="sc">~</span> <span class="fu">tibble</span>(</span>
<span id="cb9-83"><a href="#cb9-83" tabindex="-1"></a>        <span class="at">cell_id =</span> .x<span class="sc">$</span>cell_id,</span>
<span id="cb9-84"><a href="#cb9-84" tabindex="-1"></a>        <span class="at">context =</span> .x<span class="sc">$</span>context,</span>
<span id="cb9-85"><a href="#cb9-85" tabindex="-1"></a>        <span class="at">markerScore =</span> .x<span class="sc">$</span>markerScore,</span>
<span id="cb9-86"><a href="#cb9-86" tabindex="-1"></a>        <span class="at">scoreThreshold =</span> .x<span class="sc">$</span>scoreThreshold,</span>
<span id="cb9-87"><a href="#cb9-87" tabindex="-1"></a>        <span class="at">foldChange =</span> .x<span class="sc">$</span>foldChange,</span>
<span id="cb9-88"><a href="#cb9-88" tabindex="-1"></a>        <span class="at">cell_lineage =</span> .x<span class="sc">$</span>cell_lineage,</span>
<span id="cb9-89"><a href="#cb9-89" tabindex="-1"></a>        <span class="at">cell_term =</span> .x<span class="sc">$</span>cell_term</span>
<span id="cb9-90"><a href="#cb9-90" tabindex="-1"></a>      ))</span>
<span id="cb9-91"><a href="#cb9-91" tabindex="-1"></a>    <span class="fu">bind_cols</span>(gene_info, effect_sizes)</span>
<span id="cb9-92"><a href="#cb9-92" tabindex="-1"></a>  })</span>
<span id="cb9-93"><a href="#cb9-93" tabindex="-1"></a></span>
<span id="cb9-94"><a href="#cb9-94" tabindex="-1"></a><span class="co"># Visualize fold changes by cell type for a selected gene</span></span>
<span id="cb9-95"><a href="#cb9-95" tabindex="-1"></a>selected_gene <span class="ot">&lt;-</span> <span class="st">&quot;HMGB1&quot;</span></span>
<span id="cb9-96"><a href="#cb9-96" tabindex="-1"></a>df_filtered <span class="ot">&lt;-</span> df_final <span class="sc">%&gt;%</span> <span class="fu">filter</span>(symbol <span class="sc">==</span> selected_gene)</span>
<span id="cb9-97"><a href="#cb9-97" tabindex="-1"></a></span>
<span id="cb9-98"><a href="#cb9-98" tabindex="-1"></a><span class="fu">ggplot</span>(df_filtered, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">reorder</span>(cell_term, <span class="sc">-</span>foldChange), <span class="at">y =</span> foldChange)) <span class="sc">+</span></span>
<span id="cb9-99"><a href="#cb9-99" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&quot;identity&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;steelblue&quot;</span>) <span class="sc">+</span></span>
<span id="cb9-100"><a href="#cb9-100" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb9-101"><a href="#cb9-101" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb9-102"><a href="#cb9-102" tabindex="-1"></a>    <span class="at">title =</span> <span class="fu">paste</span>(<span class="st">&quot;Fold Change by Cell Type for Gene&quot;</span>, selected_gene),</span>
<span id="cb9-103"><a href="#cb9-103" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Cell Type&quot;</span>,</span>
<span id="cb9-104"><a href="#cb9-104" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Fold Change&quot;</span></span>
<span id="cb9-105"><a href="#cb9-105" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb9-106"><a href="#cb9-106" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb9-107"><a href="#cb9-107" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb9-108"><a href="#cb9-108" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb9-109"><a href="#cb9-109" tabindex="-1"></a>    <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">face =</span> <span class="st">&quot;bold&quot;</span>)</span>
<span id="cb9-110"><a href="#cb9-110" tabindex="-1"></a>  )</span></code></pre></div>

</body>
</html>
